### PR TITLE
[DRAFT][MPQEditor] Introduce MPQEditor

### PR DESCRIPTION
**Pull request by stamalakhov**
This draft introduces more fresh version of `MPQEditor` for testing with tests and screenrecordings of its capabilities.

Naming and validation of `mpq.json`:

https://user-images.githubusercontent.com/112689352/229081677-e6ca1ade-00da-4450-b81e-6c424c8456ab.mp4

Changing quantization parameters and saving:

https://user-images.githubusercontent.com/112689352/228561235-cc420b94-f249-4ab6-9506-e6269b822c43.mp4

Adding specific layers and update of actual default layers:

https://user-images.githubusercontent.com/112689352/228562694-cef536d7-e566-4511-a196-cd2e26defeb6.mp4

ONE-vscode-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: src/Tests/MPQEditor/MPQEditor.test.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+      const dirPath: string = testBuilder.dirInTemp;
+
+      const content = `
+            empty content
+            `;
+
+      let targetNames: string[] = [
+        "model.mpq.json",
+        "model(1).mpq.json",
+        "model(1)(1).mpq.json",
```
Comment (dayo09) [inline src/Tests/MPQEditor/MPQEditor.test.ts:65]: I wonder why you have decided to generate model name as such?
I think it's rather natural that when model.mpq.json and model(1).mpq.json exists, model(2).mpq.json be the next..

File: src/MPQEditor/MPQEditor.ts [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+
+  /**
+   * Add layers to edit their quantization parameters
+   */
+  private hadleAddSpecificLayerFromDialog(document: vscode.TextDocument) {
+    const nodes =
+      this._mpqDataMap[document.uri.toString()].getDefaultModelLayers();
+    const pickOptions = {
+      canPickMany: true,
+    };
```
Comment (dayo09) [inline src/MPQEditor/MPQEditor.ts:414]: How about applying those quickPickOptions? 😄 

```suggestion
    const pickOptions = {
      title: "Add specific layers to edit their quantization parameters",
      matchOnDescription: true,
      matchOnDetail: true,
      placeHolder: "Select layers",
      ignoreFocusOut: true,
      canPickMany: true,
    };
```

Comment (dayo09) [reply src/MPQEditor/MPQEditor.ts:414]: 190e428b-2d12-4309-a964-d204e1629614

File: src/Tests/MPQEditor/MPQEditor.test.ts (hunk 1 above)
Comment (stamalakhov) [reply src/Tests/MPQEditor/MPQEditor.test.ts:65]: @dayo09 
Thank you.
Please see https://github.com/Samsung/ONE-vscode/pull/1505#issuecomment-1437715625.
Just adding `(1)` is easy to implement, but i can change it to another scheme if needed.

File: src/MPQEditor/MPQEditor.ts (hunk 2 above)
Comment (dayo09) [reply src/MPQEditor/MPQEditor.ts:414]: it will
- title: show a title "Add specific~"
- placeHolder: show a place holder "Select layers"
- ignoreFocusOut: stay even if you clicks other sides
- matchOnDescription/Detail: (well it won't affect cuz we don't have description/details in the list elements)

Comment (stamalakhov) [reply src/MPQEditor/MPQEditor.ts:414]: Ahhh. Thank you very much!

File: src/MPQEditor/MPQEditor.ts [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+        dirPath,
+        baseMPQName + MPQEditorProvider.fileExtension
+      );
+      if (!fs.existsSync(mpqPath)) {
+        return baseMPQName + MPQEditorProvider.fileExtension;
+      }
+      baseMPQName = baseMPQName + "(1)";
+    }
+    return "";
+  }
```
Comment (dayo09) [inline src/MPQEditor/MPQEditor.ts:78]: This function will not work as expected when the trial repeats more than 5 times.
It will return the "", which is not expected.

On top of that, I think it's natural to have increasing numbers (1)->(2)->...
Could you revise this function?

File: src/MPQEditor/MPQEditor.ts [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+  ): Promise<vscode.Uri | undefined> {
+    const content =
+      '{"default_quantization_dtype": "uint8",' +
+      '"default_granularity": "channel",' +
+      '"layers": [],' +
+      '"model_path": ' +
+      '"' +
+      circleName +
+      '"' +
+      "}";
```
Comment (dayo09) [inline src/MPQEditor/MPQEditor.ts:93]:
```suggestion
      `{"default_quantization_dtype": "uint8",
      "default_granularity": "channel",
      "layers": [],
      "model_path": "${circleName}"`;
```

You can use this \`\` string format :-D

File: src/MPQEditor/MPQEditor.ts (hunk 3 above)
Comment (stamalakhov) [reply src/MPQEditor/MPQEditor.ts:78]: Yes. I can. Thank you.

File: src/MPQEditor/MPQEditor.ts [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+   */
+  public static createMPQJson(uri: vscode.Uri): void {
+    const dirPath = path.parse(uri.path).dir;
+    const modelName = path.parse(uri.path).name;
+    const extName = path.parse(uri.path).ext.slice(1);
+    const circleName = path.parse(uri.path).base;
+
+    // try to guess unoccupied name for mpq.json
+    let mpqName = MPQEditorProvider.findMPQName(modelName, dirPath);
+    if (mpqName.length < 1) {
```
Comment (dayo09) [inline src/MPQEditor/MPQEditor.ts:142]:
```suggestion
    if (mpqName.length == 0) {
```

I think it's more explicit to test an empty string.

File: src/MPQEditor/MPQEditor.ts (hunk 4 above)
Comment (stamalakhov) [reply src/MPQEditor/MPQEditor.ts:93]: ok. Thank you.

File: src/MPQEditor/MPQEditor.ts [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+   */
+  public static createMPQJson(uri: vscode.Uri): void {
+    const dirPath = path.parse(uri.path).dir;
+    const modelName = path.parse(uri.path).name;
+    const extName = path.parse(uri.path).ext.slice(1);
+    const circleName = path.parse(uri.path).base;
+
+    // try to guess unoccupied name for mpq.json
+    let mpqName = MPQEditorProvider.findMPQName(modelName, dirPath);
+    if (mpqName.length == 0) {
```
Comment (dayo09) [inline src/MPQEditor/MPQEditor.ts:138]:
```suggestion
    if (mpqName.length === 0) {
```
Oops sorry.

File: src/MPQEditor/MPQEditor.ts (hunk 3 above)
Comment (stamalakhov) [reply src/MPQEditor/MPQEditor.ts]: @dayo09 
I've implemented (and updated tests and draft description) simple version of it , in the form of
```
for (let i =0; ; i++)
```
do i need to constrain somehow index (by 10000 e.g.) ?
Or it was assumed to some form `glob` query , e.g.
```
1. find the longest filename with the pattern *(*).mpq.json
2. extract number from round brackets
3. return number + 1 as possible answer
```
?

File: src/extension.ts [hunk 7]
```diff
@@ -77,6 +78,13 @@ export function activate(context: vscode.ExtensionContext) {
 
   MetadataViewerProvider.register(context);
 
+  MPQEditorProvider.register(context);
+
+  Logger.info("activated");
+
+  // Set context as a global
+  (global as any).extensionContext = context;
```
Comment (dayo09) [inline src/extension.ts:86]: Q. I wonder why is this needed?

File: src/MPQEditor/MPQEditor.ts [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+  /**
+   * find unoccupied mpq file-name
+   */
+  public static findMPQName(baseMPQName: string, dirPath: string): string {
+    if (baseMPQName.length === 0) {
+      throw new Error("Invalid mixed precision quantization file name");
+    }
+    const baseName = baseMPQName;
+    let mpqName: string = "";
+    for (let i = 0; ; i++) {
```
Comment (dayo09) [inline src/MPQEditor/MPQEditor.ts:71]: How about putting a large number as a limit for safety?

Comment (stamalakhov) [reply src/MPQEditor/MPQEditor.ts:71]: > how about putting a large number as a limit for safety in the for loop?

 @dayo09
Thank you. Yes. It seems like an option, but which one? 100? 1000? ...

File: src/extension.ts (hunk 7 above)
Comment (stamalakhov) [reply src/extension.ts:86]: > Q. I wonder why is this needed?

AFAIK no. It's needed e.g. in tests to construct `MPQEditorProvider`, but it was used `constructor` only. So it can be omitted for now, because `constructor` is trivial.

File: src/MPQEditor/MPQEditor.ts (hunk 8 above)
Comment (stamalakhov) [reply src/MPQEditor/MPQEditor.ts:71]: @dayo09 
Other option: 
we may query `fs` how many `mpq.json` files are there (`max_num`), and set maximal number of iterations as `max_num+1`.

File: src/MPQEditor/MPQEditor.ts [hunk 9]
```diff
@@ -41,13 +47,148 @@ export class MPQEditorProvider implements vscode.CustomTextEditorProvider {
...
+      .showInputBox({
+        title: `Create mixed precision quantization configuration for '${modelName}.${extName}' :`,
+        placeHolder: `Enter a file name`,
+        value: mpqName,
+        valueSelection: [
+          0,
+          mpqName.length - `${MPQEditorProvider.fileExtension}`.length,
+        ],
+        validateInput: (mpqName: string): string | undefined => {
+          return MPQEditorProvider.validateMPQName(dirPath, mpqName);
```
Comment (stamalakhov) [inline src/MPQEditor/MPQEditor.ts:163]: This code uses `validateMPQName`

**Comment by stamalakhov** [comment]
Seems like the whole draft is merged for now, so we can close it.
